src/character.cpp

```diff
@@ -2904,7 +2904,7 @@ item *Character::try_add( item it, const item *avoid, const bool allow_wield )
         // this will set ret to either it, or to stack where it was placed
         pocket.second->add( it, &ret );
         if( !keep_invlet && ( !it.count_by_charges() || it.charges == ret->charges ) ) {
-            inv->update_invlet( *ret );
+            inv->update_invlet( *ret, true, avoid );
```

Comment:
`avoid` is not always the original item. For example, `item_pocket::handle_liquid_or_spill` takes `avoid` as the container to avoid, which calls `i_add_or_drop` when the content is solid, in which case `it` is the content whereas `avoid` is the container.

Reply:
I'll have to add a new parameter to that function then, but i was hoping to avoid large changes.

Reply:
While working on adding a new parameter i noticed in player.cpp:1820 player::wear it calls has_item with a stack local copy of an item. has_item just seems to do a pointer comparison which will always fail unless i am missing a global specialiced item* == operator somewhere. If it does indeed work then i could also use that which would simplify things a lot.

Reply:
I don't think there is an overload like that; that would be rather hacky and prone to confusion.

Reply:
In that case i think most checks in player:wear are broken as they are all comparing pointers to a stack local copy of the item.

Reply:
I made an issue for it #50194.

Reply:
I have added the new parameter and gone through all the call sites (so damn many) and modified where necessary. The patch is now larger than what i had hoped for but it should handle your frozen liquid case.

I placed the new parameter after the existing avoid so other patches should not break due to this. At most they need to add a nullptr parameter if they were using the trailing boolean settings.